src/Avalonia.Styling/Styling/NthChildSelector.cs

```diff
@@ -0,0 +1,134 @@
+﻿#nullable enable
+using System;
+using System.Text;
+
+using Avalonia.LogicalTree;
+
+namespace Avalonia.Styling
+{
+    public interface IChildIndexProvider
+    {
+        (int Index, int? TotalCount) GetChildIndex(ILogical child);
+    }
+
+    public class NthLastChildSelector : NthChildSelector
+    {
+        public NthLastChildSelector(Selector? previous, int step, int offset) : base(previous, step, offset, true)
+        {
+        }
+    }
+
+    public class NthChildSelector : Selector
+    {
+        private const string NthChildSelectorName = "nth-child";
+        private const string NthLastChildSelectorName = "nth-last-child";
+        private readonly Selector? _previous;
+        private readonly bool _reversed;
+
+        internal protected NthChildSelector(Selector? previous, int step, int offset, bool reversed)
+        {
+            _previous = previous;
+            Step = step;
+            Offset = offset;
+            _reversed = reversed;
+        }
+
+        public NthChildSelector(Selector? previous, int step, int offset)
+            : this(previous, step, offset, false)
+        {
+
+        }
+
+        public override bool InTemplate => _previous?.InTemplate ?? false;
+
+        public override bool IsCombinator => false;
+
+        public override Type? TargetType => _previous?.TargetType;
+
+        public int Step { get; }
+        public int Offset { get; }
+
+        protected override SelectorMatch Evaluate(IStyleable control, bool subscribe)
+        {
+            var logical = (ILogical)control;
+            var controlParent = logical.LogicalParent;
+
+            if (controlParent is IChildIndexProvider childIndexProvider)
```

Comment:
Would it be worth falling back to using the logical children of a control if it doesn't implement `IChildIndexProvider`?

Reply:
Not sure in what case it will be useful, when any Panel already implements this interface, but it shouldn't cause any problems either.

Reply:
Actually, no. Panel is from Controls project, while this one is from Styling.
We can change it after merging core libs.